### PR TITLE
#9 Added support for command in container

### PR DESCRIPTION
**Pull request by shadpro**
- #9 Support for setting command (optional) for container
- `.scalafmt` added
- Updated tests to cover the new option

### CODE REVIEW

File: .scalafmt.conf [hunk 1]
```diff
@@ -0,0 +1,4 @@
+continuationIndent.defnSite = 2
```
Comment (vaslabs) [inline .scalafmt.conf]: thanks for adding this 👍

File: src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala [hunk 2]
```diff
@@ -64,22 +63,23 @@ class KubernetesStateYamlSpec extends FlatSpec with Matchers with ScalaCheckProp
...
+          .withImage(dockerImage)
+          .withProbes(
+            livenessProbe = HttpProbe(HttpGet("/health", 8080, List.empty), initialDelay = 3 seconds, period = 5 seconds),
+            readinessProbe = NoProbe
+          )
+          .replicas(1)
+          .addPorts(List(Port(Some("app"), 8080)))
+          .annotateSpecTemplate(Map(metadataKey -> metadataValue))
+          .env(envName, envValue)
+          .addCommand(Seq("java"))
```
Comment (vaslabs) [inline src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala]: Can we have here an arbitrary for the command to test the following:
- Empty list commands
- Single word commands
- Commands with arguments

File: src/main/scala/kubeyml/deployment/api/package.scala [hunk 3]
```diff
@@ -33,65 +33,75 @@ package object api {
...
+            Map.empty
+          ),
+          TemplateSpec(
+            List(
+              Container(
+                service,
+                image,
+                List.empty,
+                List.empty,
+                IfNotPresent,
```
Comment (vaslabs) [inline src/main/scala/kubeyml/deployment/api/package.scala]: is there a reason for changing that? My initial thinking is I would like the Always to be safe, and IfNotPresent is an optimisation that you can change (with an additional sbt key).
I can open another ticket for this to give the option

Comment (vaslabs) [reply src/main/scala/kubeyml/deployment/api/package.scala]: ticket created https://github.com/vaslabs/sbt-kubeyml/issues/9

File: src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala (hunk 2 above)
Comment (vaslabs) [reply src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala]: note: https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/

Comment (vaslabs) [reply src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala]: API wise, we could make the following choices:

- Seq.head goes in command
- Seq.tail goes in arguments

Or we make something like

- addCommand(cmd: String, arguments: List[String])

And the json encoders and decoders should bridge the differences between this and kubernetes schema

then, the sbt keys can be, one for command which should be a NonEmptyString and one for arguments which defaults to an empty list

Comment (vaslabs) [reply src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala]: ok a note on the command, it doesn't have to be empty string, we just need to make this rules hold

> When you override the default Entrypoint and Cmd, these rules apply:
> 
> If you do not supply command or args for a Container, the defaults defined in the Docker image are used.
> 
> If you supply a command but no args for a Container, only the supplied command is used. The default EntryPoint and the default Cmd defined in the Docker image are ignored.
> 
> If you supply only args for a Container, the default Entrypoint defined in the Docker image is run with the args that you supplied.
> 
> If you supply a command and args, the default Entrypoint and the default Cmd defined in the Docker image are ignored. Your command is run with your args.
>

File: src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala [hunk 4]
```diff
@@ -128,37 +130,46 @@ class KubernetesStateYamlSpec extends FlatSpec with Matchers with ScalaCheckProp
                |          env:
                |            - name: "${envName}"
                |              value: "${envValue}"
-               |""".stripMargin).right.get
-          if (expectedYaml != deployment.asJson) {
-            println(expectedYaml.noSpaces)
+               |""".stripMargin)
+
+        val someYaml = expectedYaml match {
```
Comment (vaslabs) [inline src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala]: Not sure what's going on here, please rebase with master and if you could try to follow the property testing paradigm. This class is only for sanitation, could you create your own to test your changes please?

Comment (vaslabs) [reply src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala]: Ah ok, I think partly is my fault for not filtering out the rule on codacy for this.
But an unsafe right.get is acceptable in a test (for me) if we avoid more complex code.
In any case I have re-written this, sorry for the confusion

Comment (shadpro) [reply src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala]: Cool! I'll have a look.

File: src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala (hunk 2 above)
Comment (shadpro) [reply src/test/scala/kubeyml/deployment/KubernetesStateYamlSpec.scala]: Preferable approach would be- Option[NonEmptyString]

File: src/main/scala/kubeyml/deployment/api/package.scala (hunk 3 above)
Comment (shadpro) [reply src/main/scala/kubeyml/deployment/api/package.scala]: Reason for changing that was by default, the images we build, our immutable. So we always use "IfNotPresent". Also during development/testing phase in local, I would prefer to use the locally image build, over the published image in docker repo.

We can have that option of choosing the defaults.

Comment (vaslabs) [reply src/main/scala/kubeyml/deployment/api/package.scala]: Ok, I have made the changes to make it configurable anyway, please remember to update the readme too

**Comment by vaslabs** [comment]
Thanks for the contribution. I'll prepare a release once I'm back from holidays, probably Monday

**Pull request merged**